pkg/validator/validators_test.go

```diff
@@ -786,6 +786,181 @@ func TestIngressBackendValidator(t *testing.T) {
+			expErrStr: "",
+		},
+		{
+			name: "invalid - source as not ServiceAccount",
+			input: &admissionv1.AdmissionRequest{
+				Kind: metav1.GroupVersionKind{
+					Group:   "v1alpha1",
+					Version: "policy.openservicemesh.io",
+					Kind:    "Retry",
+				},
+				Object: runtime.RawExtension{
+					Raw: []byte(`
+					{
+						"apiVersion": "v1alpha1",
+						"kind": "Retry",
+						"spec": {
+							"source": {
+								"kind": "Service",
+								"name": "test",
+								"namespace": "testns"
+							},
+							"destinations": [
+								{
+									"kind": "Service",
+									"name": "testdest",
+									"namespace": "testns"
+								}
+							],
+							"retryPolicy": {
+								"retryOn": "5xx"
+							}
+						}
+					}
+					`),
+				},
+			},
+			expResp:   nil,
+			expErrStr: "Retry source test in namespace testns is a Service - must be a ServiceAccount",
+		},
+		{
+			name: "invalid destination as not Service",
+			input: &admissionv1.AdmissionRequest{
+				Kind: metav1.GroupVersionKind{
+					Group:   "v1alpha1",
+					Version: "policy.openservicemesh.io",
+					Kind:    "Retry",
+				},
+				Object: runtime.RawExtension{
+					Raw: []byte(`
+					{
+						"apiVersion": "v1alpha1",
+						"kind": "Retry",
+						"spec": {
+							"source": {
+								"kind": "ServiceAccount",
+								"name": "test",
+								"namespace": "testns"
+							},
+							"destinations": [
+								{
+									"kind": "ServiceAccount",
+									"name": "testdest",
+									"namespace": "testns"
+								}
+							],
+							"retryPolicy": {
+								"retryOn": "5xx"
+							}
+						}
+					}
+					`),
+				},
+			},
+			expResp:   nil,
+			expErrStr: "Retry destination testdest in namespace testns is a ServiceAccount - must be a Service",
+		},
+		{
+			name: "invalid duplicate destination",
+			input: &admissionv1.AdmissionRequest{
+				Kind: metav1.GroupVersionKind{
+					Group:   "v1alpha1",
+					Version: "policy.openservicemesh.io",
+					Kind:    "Retry",
+				},
+				Object: runtime.RawExtension{
+					Raw: []byte(`
+					{
+						"apiVersion": "v1alpha1",
+						"kind": "Retry",
+						"spec": {
+							"source": {
+								"kind": "ServiceAccount",
+								"name": "test",
+								"namespace": "testns"
+							},
+							"destinations": [
+								{
+									"kind": "Service",
+									"name": "testdest",
+									"namespace": "testns"
+								},
+								{
+									"kind": "Service",
+									"name": "testdest",
+									"namespace": "testns"
+								}
+							],
+							"retryPolicy": {
+								"retryOn": "5xx"
+							}
+						}
+					}
+					`),
+				},
+			},
+			expResp:   nil,
+			expErrStr: "Duplicate destinations - repeated service testdest in namespace testns",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			assert := tassert.New(t)
+
+			resp, err := retryValidator(tc.input)
+			assert.Equal(tc.expResp, resp)
+			if err != nil {
+				assert.Equal(tc.expErrStr, err.Error())
+			}
```

Comment:
This could pass when (I would argue) it should fail when the test defines a non-empty `tc.expErrStr` but a nil error is actually returned. One way to get around that could be with something like:
```suggestion
			errStr := ""
			if err != nil {
				errStr = err.Error()
			}
			assert.Equal(tc.expErrStr, errStr)
```